apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java

```diff
@@ -234,14 +233,31 @@ void onConnectionError() {
             } catch (Exception e) {
                 logger.debug("Exception while interacting with APM Server", e);
             } finally {
-                if (connection != null) {
-                    connection.disconnect();
-                }
+                closeInputStream(connection);
             }
         }
         return results;
     }
 
+    private void closeInputStream(@Nullable HttpURLConnection connection) {
+        if (connection != null) {
+            try {
+                int responseCode = connection.getResponseCode();
+                InputStream responseInputStream = null;
+                if (responseCode < 400) {
+                    responseInputStream = connection.getInputStream();
+                } else if (responseCode >= 400) {
+                    responseInputStream = connection.getErrorStream();
+                }
+                if (responseInputStream != null) {
+                    responseInputStream.close();
```

Comment:
We not only have to close the input stream but also have to make sure to consume the stream. Use `org.stagemonitor.util.IOUtils#consumeAndClose` for that.
However, I'm not sure what happens if trying to consume an already closed stream.

Reply:
Inside  ```IOUtils#consumeAndClose``` exist exception handling:
```
	public static void consumeAndClose(InputStream is) {
		if (is == null) {
			return;
		}
		try {
			while (is.read() != EOF) {}
		} catch (IOException e) {
			logger.warn(e.getMessage(), e);
		} finally {
			closeQuietly(is);
		}
	}

```

Reply:
Do you see an issue with that?

Reply:
No, there are no issues.